libcaf_core/src/actor_system_config.cpp

```diff
@@ -332,20 +347,52 @@ error actor_system_config::parse(string_list args, std::istream& config) {
   return none;
 }
 
+error actor_system_config::parse(string_list args) {
+  if (auto&& [err, path] = extract_config_file_path(args); !err) {
+    std::ifstream conf;
+    // No error but an empty path means --config-file=ARG was missing.
+    if (!path.empty()) {
+      conf.open(path);
+    } else {
+      // Try config_file_path and if that fails try the alternative paths.
+      auto try_open = [this, &conf](const auto& what) {
+        if (what.empty())
+          return false;
+        conf.open(what);
+        if (conf.is_open()) {
+          set("global.config-file", what);
+          return true;
+        } else {
+          return false;
+        }
+      };
+      try_open(config_file_path)
+        || std::any_of(config_file_path_alternatives.begin(),
+                       config_file_path_alternatives.end(), try_open);
+    }
+    return parse(std::move(args), conf);
+  } else {
+    return err;
+  }
+}
+
 error actor_system_config::parse(string_list args,
                                  const char* config_file_cstr) {
-  // Override default config file name if set by user.
-  if (config_file_cstr != nullptr)
-    config_file_path = config_file_cstr;
-  // CLI arguments always win.
-  if (auto err = extract_config_file_path(args))
-    return err;
-  if (config_file_path == "$DEFAULT") {
-    std::ifstream conf{"caf-application.conf"};
+  if (config_file_cstr == nullptr) {
+    return parse(std::move(args));
+  } else if (auto&& [err, path] = extract_config_file_path(args); !err) {
+    std::ifstream conf;
+    if (!path.empty()) {
+      conf.open(path);
+    } else {
+      conf.open(config_file_cstr);
+      if (conf.is_open())
+        set("global.config-file", config_file_cstr);
+    }
     return parse(std::move(args), conf);
```

Comment:
If `extract_config_file_path` fails with an error, could it happen that `config_file_cstr` exists, is readable, and ... should be read?

Reply:
No. If `extract_config_file_path` fails then we should always stop parsing.

> do we have a test case for `actor_system_config::extract_config_file_path`

This is a private function, so we currently can't access it in isolation. Making it public only for the sake of testing does not strike me as reasonable, because someone always starts depending on public functions eventually. Ripping it out of the class and moving it into the `detail` namespace might work, but the member function accesses three fields of the class. So there would be quite a bit of refactoring necessary and the function also has assumptions (like the name of the config option) that don't really make sense for a free function. Testing the `parse` functions would be the solution, but that function does file I/O. So until we have a setup for integration tests in CAF, I think we need to test changes like this by hand. Fortunately we don't change that often.